Enable time-dependent effective areas determination in iris_tools

kakirastern opened the pull request:
Picks up from PR #102 and addresses issue #27.

Goal
====
To allow the time-dependent effective areas to be derived using get_iris_response() in iris_tools in the same way that it is done in SSW.

Expected Outcomes
===============
1. A function for deriving the time-dependent IRIS response function.
2. Benchmarking and unit tests so this new software can be reliably maintained.
3. Updated intensity conversion methods between instrument and physical units that correct for the time observations were taken.

Project Description
===================
The Interface Region Imaging Spectrograph (IRIS) is a NASA Small Explorer satellite designed to make spectroscopic and imaging observations of the solar chromosphere and transition region. An important tool in interpreting these observations is the instrument response function which expresses the relationship between the physical intensity of light entering the instrument and the data number (DN) recorded by the CCD detectors. The response function was carefully measured by the IRIS team before launch and this is currently available to IRISpy users for their analysis.

However, due to time-dependent factors like degradation, it is known that the response function evolves with time in orbit. For this reason the IRIS team has developed a fitting algorithm based on calibration flights that predicts the response function as a function of time. A direct translation of this algorithm from its original language, IDL (Interactive Data Langauage), into Python has already been performed but requires more work before it can be merged into IRISpy.

The `.geny` files needed can be accessed at [https://sohowww.nascom.nasa.gov/solarsoft/iris/response/](https://sohowww.nascom.nasa.gov/solarsoft/iris/response/). 

References:
[iris_get_response.pro](https://sohowww.nascom.nasa.gov/solarsoft/iris/idl/lmsal/util/iris_get_response.pro)
[fit_iris_xput.pro](https://www.heliodocs.com/xdoc/xdoc_print.php?file=$SSW/iris/idl/lmsal/calibration/fit_iris_xput.pro)

Roadmap - fit_iris_xput
=================
Triaging
--------
* Figure out how to run the function without crashing assuming, this is possible.
    * If the function cannot be made to run, are there simple bug fixes that can address this or is deeper understanding of what the function is doing required?
* Use print statements or other techniques to go through the code and understand what each variable represents, or at the very least what type it is.  It may be helpful to look at the docstring and comments of the IDL version of the code as part of this.  Write up a document that explains this, perhaps on the IRISpy wiki.
* From the above step, write a preliminary docstring for ```fit_iris_xput``` explaining what the inputs and outputs.

Benchmarking
---------------
* Write a test function that will allow us to test and understand the output.
* Coordinate efforts to compare results from the IDL and IRISpy versions to verify they are giving the same answers or where differences are occurring.  (Dan have the IDL version installed.)
* From this, write up required To-Do tasks needed to fix any discrepancies.

Bug-fixing
-----------
Make code fixes to get the Python version giving the same or effectively the same answers as the IDL version.
* Most common type of bugs encountered comes from syntax, especially from converting between programming languages. 
* One of the more glaring and easily omitted error encountered stemmed from erroneous indentation. 

Improving Code Pythonically
----------------------------
Once we have a function that returns the same or similar values as the IDL version, the code will have to be made more Pythonic as well as updated to reflect more recent changes in SunPy practices.  Benchmarking/testing must be repeated and updated at each stage to ensure we are not changing the values of the output.
* Using AstroPy's [`download_file`](http://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html) instead of Requests' [`get`](https://2.python-requests.org/en/master/user/quickstart/#make-a-request) to download the `.geny` files. This adds the `cache` and `timeout` functionalities to the download process.
* Converting the `time_obs` and `time_cal_coeffs` into `astropy.time.time` objects before subtracting to obtain a time difference called `t_diff` (to be converted into years), which is used to compute the final fit.

Incorporating Time-dependent Fitted Response into get_iris_response
----------------------------------------------------------------------
* Modify the `get_iris_response` function in iris_tools.py to incorporate consideration of time-dependency. 
* [_get_interpolated_effective_area](https://github.com/sunpy/irispy/pull/108/files#diff-dee23d5475a6ed5e368938455c92fb2eL681) must include a call to ```fit_iris_xput``` and add a time input.
* Add a docstring to the `_get_interpolated_effective_area` function. 
* Tests of above functions must include cases where a fitted effective area is used.

Benchmarking Cases from SSW
=======================
The .geny files can be read in Python using scipy's read save function.  See [here](https://github.com/sunpy/irispy/blob/master/irispy/iris_tools.py#L156) for an example.  Also note that the out is converted to a ```dict``` for easier handling the line after.  The same attribute names should be maintained between the the IDL and Python structures.

1: Result for a start time of period of constant coef/cal_coef
------------------------------------------------------------
Lines starting with ```IDL>``` are input
```;``` is the IDL comment symbol
Lines without ```IDL>``` at the start are output.
```idl
IDL> restore, 'iris_sra_c_20161022.geny', /ver ; Restore 2016 response function file. Structure read out in to variable, p0.  For Python version, read in file using scipy's readsav function.  See https://github.com/sunpy/irispy/blob/master/irispy/iris_tools.py#L156
% RESTORE: Portable (XDR) SAVE/RESTORE file.
% RESTORE: Save file written by wuelser@berna, Thu Dec  8 15:50:09 2016.
% RESTORE: IDL version 8.5 (darwin, x86_64).
% RESTORE: Restored variable: P0.
IDL> time_cal_coeffs = p0.c_f_time ; Define FUV time_cal_coeffs input from the p0 variable
IDL> print, time_cal_coeffs
  1.09417e+09  1.10186e+09
  1.10532e+09  1.12933e+09
  1.12951e+09  1.13201e+09
  1.13642e+09  1.16182e+09
  1.16199e+09  2.24052e+09

IDL> ; Define FUV cal_coeffs input from the p0 variable
IDL> cal_coeffs0 = p0.coeffs_fuv[*, *, 0]
IDL> print, cal_coeffs0
     0.523856     0.274797     -6.45654
     0.236928     0.395193     -3.80189
     0.411110    0.0621537     -24.5471
     0.228041     0.247113     -1.99526
     0.128971     0.198371    -0.891251

IDL> cal_coeffs1 = p0.coeffs_fuv[*, *, 1]
IDL> print, cal_coeffs1
    -0.654571      2.90411    -0.891251
     0.957263     0.792720     -7.58577
     0.209228     0.834837     -1.94984
     0.825466     0.156275     -2.69153
     0.675337     0.149698     -1.65959

IDL> cal_coeffs2 = p0.coeffs_fuv[*, *, 2]
IDL> print, cal_coeffs2
    -0.654571      2.90411    -0.891251
     0.957263     0.792720     -7.58577
     0.209228     0.834837     -1.94984
     0.825466     0.156275     -2.69153
     0.675337     0.149698     -1.65959

IDL> ; Define time(s)
IDL> tt = fltarr(1)
IDL> tt[0] = 1.0941696e+09 ; tt is equivalent to time_obs in Python version.

IDL> ; Perform fits for FUV ranges.
IDL> fit0_fuv = fit_iris_xput(tt, p0.c_f_time, p0.coeffs_fuv[*, *, 0])
IDL> fit1_fuv = fit_iris_xput(tt, p0.c_f_time, p0.coeffs_fuv[*, *, 1])
IDL> fit2_fuv = fit_iris_xput(tt, p0.c_f_time, p0.coeffs_fuv[*, *, 2])
IDL> print, fit0_fuv, fit1_fuv, fit2_fuv
0.79865301
2.2495413
2.2495413
```

IDL> ; Perform fits for NUV ranges. Note the changes in p0.c_n_time and p0.coeffs_nuv variable names.
IDL> fit0_nuv = fit_iris_xput(tt, p0.c_n_time, p0.coeffs_nuv[*, *, 0])
IDL> fit1_nuv = fit_iris_xput(tt, p0.c_n_time, p0.coeffs_nuv[*, *, 1])
IDL> fit2_nuv = fit_iris_xput(tt, p0.c_n_time, p0.coeffs_nuv[*, *, 2])
IDL> print, fit0_nuv, fit1_nuv, fit2_nuv
0.23529011
0.25203046
0.25265095


Incorporating Time-dependent Fitted Response into Wider IRISpy Code-base
========================================================
* Functions further up the chain must also include a time input and pass that on to ```_get_interpolated_effective_area```.  This chain includes ```iris_tools.calculate_photons_per_sec_to_radiance_factor```, ```iris_tools.convert_or_undo_photons_per_sec_to_radiance```,  and```IRISSpectrogramCube.convert_to```.  Must confirm if there are more cases where this chain leads.
* Tests of above functions must include cases where a fitted effective area is used.


Final Pythonic Tidy Up
================
Go through code produced above and make any cumbersome bits of code more Pythonic.  This may also involve tidying up or writing a few new tests.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -21,6 +21,9 @@
 from sunpy.util.net import check_download_file
 from ndcube import NDCube
 
+from astropy.time import Time
```
Move this line among the other astropy imports above.

kakirastern replied:
Yup, moved `from astropy.time import Time` to above along with the other astropy imports.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -21,6 +21,9 @@
 from ndcube import NDCube
 
+from astropy.time import Time
+from IPython.core import debugger ; debug = debugger.Pdb().set_trace
```
Delete this import.

kakirastern replied:
Yup, deleted `from IPython.core import debugger ; debug = debugger.Pdb().set_trace`.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -211,10 +221,71 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
 @custom_model
 def _gaussian1d_on_linear_bg(x, amplitude=None, mean=None, standard_deviation=None,
-                             constant_term=None, linear_term=None):
+                                 constant_term=None, linear_term=None):
```
Undo these extra leading spaces.

kakirastern replied:
Done, extra leading spaces deleted.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -171,32 +174,39 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        #     for j, t in enumerate(row):
+        #         c_f_time[i][j] = parse_time(float(t))
+        # iris_response["C_F_TIME"] = c_f_time
+        c_f_time = iris_response["C_F_TIME"]
```
Delete this line and uncomment the above paragraph.

kakirastern replied:
Done.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -171,32 +174,39 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        #     for j, t in enumerate(row):
+        #         c_n_time[i][j] = parse_time(float(t))
+        # iris_response["C_N_TIME"] = c_n_time
+        c_n_time = iris_response["C_N_TIME"]
```
Delete this line and uncomment the above paragraph.

kakirastern replied:
Done.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -171,32 +174,39 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        #         for k, t in enumerate(column):
+        #             c_s_time[i][j][k] = parse_time(float(t))
+        # iris_response["C_S_TIME"] = c_s_time
+        c_s_time = iris_response["C_S_TIME"]
```
Delete this line and uncomment the above paragraph.

kakirastern replied:
Done.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -211,10 +221,71 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
     return iris_response
 
+def fit_iris_xput(time_obs, time_cal_coeffs, cal_coeffs):
 
```
A docstring is required here explaining what the inputs and outputs are.  This will take some investigation.

kakirastern replied:
Added a draft docstring in. Please review to check for correctness.

kakirastern replied:
Added a draft docstring that needs reviewing to ascertain correctness.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -211,10 +221,71 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+    if size_time_cal_coeffs[1] != 2 or size_cal_coeffs[1]  < 2:
+
+       print('ERROR: Incorrect number of elements either in time_cal_coeffs or in cal_coeffs.')
+       return
```
See comment on this line in PR #102.

kakirastern replied:
Edited as per suggestion given in PR #102.

DanRyanIrish commented:
Hi @kakirastern.  Thanks for setting up this PR!  I've left some comments outlining some simple initial changes.

I think there are a few things we could do to help us understand what we need to do.  I've laid those out in the description of the PR above for easy reference.  Let me know here what you think.

kakirastern commented:
You are welcome @DanRyanIrish! I really like the look of the PR and think the plan appears very sound... Looking forward to contributing more to it soon. I have made some preliminary changes as per the comments in PR #102.

DanRyanIrish commented:
They look good!  Let me know whenever you're ready to report on whether you were able to get ```fit_iris_xput``` running and what bugs you found.

kakirastern commented:
Yup, sure. Will take a while so will get back to you at least a couple days later.

kakirastern commented:
Hi @DanRyanIrish! So I have managed to use `flake8` to check on `fit_iris_xput` and get it running/working on a functional level with a new commit 9215373, though I am not sure what meaningful input to feed into it to debug it rigorously. I am wondering if you have some good example I could use for debugging against?

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,88 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        # Raise ValueError as time coefficient have the wrong format.
+        raise ValueError("Incorrect number of elements either in time_cal_coeffs or in cal_coeffs.")
+        # Exponent for transition between exp.decay intervals.
+        transition_exp = 1.5
```
All lines from this one and below should not be unindented four spaces to take them out of the above ```if``` statement part of this ```if``` statement.  As it is, the code from here on will never be executed.

DanRyanIrish commented:
> Hi @DanRyanIrish! So I have managed to use `flake8` to check on `fit_iris_xput` and get it running/working on a functional level with a new commit [9215373](https://github.com/sunpy/irispy/commit/9215373a91ad5d42c02c5b5c68fde31f45c016aa), though I am not sure what meaningful input to feed into it to debug it rigorously. I am wondering if you have some good example I could use for debugging against?

I'll figure out what the meaningful input should be and run the SSW version to get the expected output.

DanRyanIrish commented:
Hi @kakirastern.  I've added a section to the description of this issue above for benchmarking cases listing the result from the SSWIDL software.  You should be able to compare the output of the Python version with this to start with.

But first see the comment I left about the indentations as nothing will work properly until that is resolved.

kakirastern commented:
Thanks @DanRyanIrish for the suggestions and info! I did run into some bugs regarding the shape of a few variables defined as numpy arrays... which will need some work to fix. From what I understand of the situation this`fit_iris_xput` is basically a rendering of some least-squares fit? In that case it might be easy to find out what's wrong and debug.

kakirastern commented:
I will focus my time and effort to completing [https://github.com/sunpy/ndcube/pull/151](https://github.com/sunpy/ndcube/pull/151) first for the time being and get back to this soon after that. Hope this arrangement is okay with you.

DanRyanIrish commented:
Sounds like a good plan @kakirastern

kakirastern commented:
So I finally was able to get the new function `fit_iris_xput` working... and I did the following:

```
>>> import numpy as np
>>> import scipy.io

>>> from irispy.iris_tools import fit_iris_xput

>>> response_file = '/Users/koychui/irispy-git/irispy/data/iris_sra_c_20161022.geny'
>>> raw_response_data = scipy.io.readsav(response_file)
>>> iris_response = dict([(name, raw_response_data["p0"][name][0]) for name in raw_response_data["p0"].dtype.names])

>>> time_obs = np.array([1.0941696e+09])
>>> time_cal_coeffs = iris_response.get('C_F_TIME')
>>> cal_coeffs0 = iris_response.get('COEFFS_FUV')[0,:,:]
>>> cal_coeffs1 = iris_response.get('COEFFS_FUV')[1,:,:]
>>> cal_coeffs2 = iris_response.get('COEFFS_FUV')[2,:,:]

>>> fit_iris_xput(time_obs, time_cal_coeffs, cal_coeffs0)
array([0.79865301])
>>> fit_iris_xput(time_obs, time_cal_coeffs, cal_coeffs1)
array([2.24954128])
>>> fit_iris_xput(time_obs, time_cal_coeffs, cal_coeffs2)
array([2.24954128])
```

If instead I perform fits for NUV ranges, I'd get the values `array([0.23529011])`, `array([0.25203046])`, and `array([0.25265094])`, with the last value deviating/smaller by `0.00000001`. 

I am not sure why the last fit carried out with the Python code yields a result that is different from the IDL fit. @DanRyanIrish, do you reckon this would imply some significant differences (in the long run)?

kakirastern commented:
@DanRyanIrish Any suggestions as to what I should do next?

DanRyanIrish commented:
That's excellent news @kakirastern!  Well done!  A different of ```0.00000001``` is insignificant.  It's probably a rounding error in the computer itself.

The next thing to do is formalize this into a test, and to do it using the ```parameterize``` decorator so we can add more cases later.  Once that's working, we should look into trying more times so we adequately cover the parameter space.

We're moving onto phase two of the roadmap already!

(By the way, next time you compare values like this can you put the IDL and Python values side-by-side so I can easily see what the differences are?  Thanks!)

kakirastern commented:
Yup, sure @DanRyanIrish. I will do better data presentation next time. (Should have made a table for it.) And, onto testing now...

kakirastern commented:
And, I will set the only error of `0.00000001` as the tolerance for the test then.

DanRyanIrish commented:
> And, I will set the only error of `0.00000001` as the tolerance for the test then.

```1e-6``` Is loads.  Probably too tight but we can loosen it as needed when we see results from other times later.

kakirastern commented:
Used the default 6 decimal places as the tolerance for the `fit_iris_xput` test in the end... 

Now I see the following failure thrown:
```
=================================== FAILURES ===================================
__________________________ test_fits_data_comparison ___________________________

iris_l2_test_raster = <iris.IRISSpectrograph instance
OBS ID: 3690215148
OBS Description: Large sparse 32-step raster 31x120 32s  C II   Si ... 1336 : [1, 32, 196, 514]    pix
     Si IV 1394 : [1, 32, 196, 510]    pix
   Mg II k 2796 : [1, 32, 196, 508]    pix>

    def test_fits_data_comparison(iris_l2_test_raster):
        """Make sure the data is the same in pyfits and irispy"""
        hdulist = fits.open(os.path.join(
            testpath, 'iris_l2_20170222_153635_3690215148_raster_t000_r00000.fits'))
        spectral_window1 = hdulist[0].header["TDESC1"]
        spectral_window2 = hdulist[0].header["TDESC2"]
        spectral_window3 = hdulist[0].header["TDESC3"]
    
        data1 = copy.deepcopy(hdulist[1].data)
        data2 = copy.deepcopy(hdulist[2].data)
        data3 = copy.deepcopy(hdulist[3].data)
    
        data1[hdulist[1].data == -200.] = np.nan
        data2[hdulist[2].data == -200.] = np.nan
        data3[hdulist[3].data == -200.] = np.nan
    
        np.testing.assert_array_almost_equal(
>           iris_l2_test_raster.data[spectral_window1].data[0].data, data1)

irispy/tests/test_spectrograph.py:145: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

x = array([[[-200., -200., -200., ..., -200., -200., -200.],
        [-200., -200., -200., ..., -200., -200., -200.],
    ...00., -200., -200., ..., -200., -200., -200.],
        [-200., -200., -200., ..., -200., -200., -200.]]], dtype=float32)
y = array([[[nan, nan, nan, ..., nan, nan, nan],
        [nan, nan, nan, ..., nan, nan, nan],
        [nan, nan, nan, ...,..., nan, nan],
        [nan, nan, nan, ..., nan, nan, nan],
        [nan, nan, nan, ..., nan, nan, nan]]], dtype=float32)
func = <ufunc 'isnan'>, hasval = 'nan'

    def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
        """Handling nan/inf.
    
        Combine results of running func on x and y, checking that they are True
        at the same locations.
    
        """
        x_id = func(x)
        y_id = func(y)
        # We include work-arounds here to handle three types of slightly
        # pathological ndarray subclasses:
        # (1) all() on `masked` array scalars can return masked arrays, so we
        #     use != True
        # (2) __eq__ on some ndarray subclasses returns Python booleans
        #     instead of element-wise comparisons, so we cast to bool_() and
        #     use isinstance(..., bool) checks
        # (3) subclasses with bare-bones __array_function__ implemenations may
        #     not implement np.all(), so favor using the .all() method
        # We are not committed to supporting such subclasses, but it's nice to
        # support them if possible.
        if bool_(x_id == y_id).all() != True:
            msg = build_err_msg([x, y],
                                err_msg + '\nx and y %s location mismatch:'
                                % (hasval), verbose=verbose, header=header,
                                names=('x', 'y'), precision=precision)
>           raise AssertionError(msg)
E           AssertionError: 
E           Arrays are not almost equal to 6 decimals
E           
E           x and y nan location mismatch:
E            x: array([[[-200., -200., -200., ..., -200., -200., -200.],
E                   [-200., -200., -200., ..., -200., -200., -200.],
E                   [-200., -200., -200., ..., -200., -200., -200.],...
E            y: array([[[nan, nan, nan, ..., nan, nan, nan],
E                   [nan, nan, nan, ..., nan, nan, nan],
E                   [nan, nan, nan, ..., nan, nan, nan],...

/anaconda3/envs/irispy-dev/lib/python3.6/site-packages/numpy/testing/_private/utils.py:733: AssertionError
```

kakirastern commented:
So I found out there are 3 lines in the file "test_spectrograph.py" which were causing problems before I commented them out. Not sure why they were there in the first place. Could have been used for debugging earlier. They actually appeared out of place before: 

[https://github.com/sunpy/irispy/pull/108/files#diff-7c42eb0d015fb3346007f43ca206dd1aR140](https://github.com/sunpy/irispy/pull/108/files#diff-7c42eb0d015fb3346007f43ca206dd1aR140)

DanRyanIrish commented:
Well done on getting the tests to pass!  Looks great.  The next step is to make the code in ```fit_iris_xput``` more Pythonic without changing the output.  I can go through the code an make suggestions on how to go about that.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -169,15 +173,17 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
         try:
             iris_response["DATE_OBS"] = parse_time(iris_response["DATE_OBS"])
-        except:
+        except Exception:
```
I think a simple ```except:``` should be fine here.

kakirastern commented:
That would be great!

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,88 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        if isinstance(cal_coeffs, np.ndarray):
+            size_time_cal_coeffs = time_cal_coeffs.shape
+            size_cal_coeffs = cal_coeffs.shape
+
```
Convert ```time_obs``` to an array of ```datetime.datetime``` objects near the start of the function.  The translations to do this are [here](https://github.com/sunpy/irispy/pull/108/files#diff-dee23d5475a6ed5e368938455c92fb2eR266).  Then later we can make the input take the form of ```datetime.datetime``` or ```astropy.time.Time``` without messing up the output.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,88 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+    fit_out = np.zeros(len(list(time_obs)), dtype=np.float64)
+
+    # Some time transformations.
+    t_cal_coeffs = np.float64(time_cal_coeffs)
```
Instead of converting to floats, convert to ```datetime.datetime``` objects.  At first glance you should be able to use the same translations used for ```time_obs```.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,88 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        # Looking for the closest time in the calibration time intervals.
+        # Differences are given in years
+        t_diff = t_obs - t_cal_coeffs
+        t_diff = t_diff/yr2sec
```
Since after above changes, ```t_obs``` and ```t_cal_coeffs``` will be ```datetime.datetime``` objects, this conversion can be removed.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,88 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        # Differences are given in years
+        t_diff = t_obs - t_cal_coeffs
+        t_diff = t_diff/yr2sec
+        t_diff = np.reshape(t_diff, size_time_cal_coeffs[0]*size_time_cal_coeffs[1])
```
I think this line is the same as ```t_diff.flatten()```.  But make sure to check.  If so, use ```.flatten()``` instead.

kakirastern replied:
Not quite, they are not the same... so abandoned this suggestion.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,88 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        t_diff = t_diff/yr2sec
+        t_diff = np.reshape(t_diff, size_time_cal_coeffs[0]*size_time_cal_coeffs[1])
+        w = (np.where((t_diff - np.abs(t_diff)) != 0))
+        w = w[0][0]
```
I think these two lines are ```w = np.where(t_diff < 0)[0][0]```.  If so use that instead.  But be sure to confirm.

kakirastern replied:
Yes, they are indeed equivalent...

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,88 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+    for i in time_obs:
+        aux_cal_coeffs = np.zeros(2*size_time_cal_coeffs[0])
+        t_obs = parse_time(i, utime=1) - t_ref
+        t_obs = t_obs.total_seconds()
```
Since ```time_obs``` and ```t_cal_coeffs``` will have been converted to ```datetime.datetime``` the two lines above can be deleted and [L271](https://github.com/sunpy/irispy/pull/108/files#diff-dee23d5475a6ed5e368938455c92fb2eR271) can be changed to ```t_diff = (t_obs - t_cal_coeffs).total_seconds()```

kakirastern commented:
Thanks @DanRyanIrish for the suggestions! Will definitely work on them over the next few days.

kakirastern commented:
Hi @DanRyanIrish, I have incorporated all of the suggested changes to render the `fit_iris_xput` function more pythonic, except for the ones that don't work in our case. Please have a look and let me know what steps to take next.

kakirastern commented:
So if I do the following:
```
import os
import pytest

import numpy as np
import numpy.testing as np_test
import astropy.units as u
from astropy.utils.data import download_file
import scipy.io
import requests

import irispy.iris_tools as iris_tools
from irispy.iris_tools import fit_iris_xput

response_url = 'https://sohowww.nascom.nasa.gov/solarsoft/iris/response/iris_sra_c_20161022.geny'
r = download_file(response_url)
raw_response_data = scipy.io.readsav(r)
iris_response = dict([(name, raw_response_data["p0"][name][0]) for name in raw_response_data["p0"].dtype.names])
time_obs = np.array([1.0941696e+09])
time_cal_coeffs0 = iris_response.get('C_F_TIME')
time_cal_coeffs1 = iris_response.get('C_N_TIME')
cal_coeffs0 = iris_response.get('COEFFS_FUV')[0,:,:]
cal_coeffs1 = iris_response.get('COEFFS_FUV')[1,:,:]
cal_coeffs2 = iris_response.get('COEFFS_FUV')[2,:,:]
cal_coeffs3 = iris_response.get('COEFFS_NUV')[0,:,:]
cal_coeffs4 = iris_response.get('COEFFS_NUV')[1,:,:]
cal_coeffs5 = iris_response.get('COEFFS_NUV')[2,:,:]
iris_fit_expected0 = np.array([0.79865301])
iris_fit_expected1 = np.array([2.2495413])
iris_fit_expected2 = np.array([2.2495413])
iris_fit_expected3 = np.array([0.23529011])
iris_fit_expected4 = np.array([0.25203046])
iris_fit_expected5 = np.array([0.25265095])
```
I would get the following
```
Case    IDL results    Pythonic results
====    ===========    ================
FUV1    0.79865301     0.79865301
FUV2    2.2495413      2.24954128
FUV3    2.2495413      2.24954128
NUV1    0.23529011     0.25605348
NUV2    0.25203046     0.2742711
NUV3    0.25265095     0.27494633

```
**Question**: Should I change the tolerance for the comparisons so that the tests will pass, or should I re-examine the code to try and debug, if it is a bug at all?

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,97 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+    # Exponent for transition between exp.decay intervals.
+    transition_exp = 1.5
+
+    fit_out = np.zeros(len(list(time_obs)), dtype=np.float64)
```
```time_obs``` shouldn't have to be converted to get to ```len```.

kakirastern replied:
No, wouldn't work since:
```
In [3]: fit_iris_xput(time_obs, time_cal_coeffs0, cal_coeffs0)                  
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-574283a177a7> in <module>
----> 1 fit_iris_xput(time_obs, time_cal_coeffs0, cal_coeffs0)

~/irispy-git/irispy/iris_tools.py in fit_iris_xput(time_obs, time_cal_coeffs, cal_coeffs)
    272     transition_exp = 1.5
    273 
--> 274     fit_out = np.zeros(list(time_obs), dtype=np.float64)
    275 
    276     # Looking for the closest time in the calibration time intervals.

TypeError: 'datetime.datetime' object cannot be interpreted as an integer
```

kakirastern commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,97 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        tmp = parse_time(float(element), utime=1) - t_ref
+        t_element = tmp.total_seconds()
+        t_cal_coeffs.append(t_element)
+    t_cal_coeffs = np.array(t_cal_coeffs).reshape(5, 2)
```
Yup, sure

kakirastern replied:
Yup, agreed

kakirastern commented:
Weird, I still got the same data...

kakirastern commented on irispy/iris_tools.py:
```diff
@@ -243,16 +243,23 @@ def fit_iris_xput(time_obs, time_cal_coeffs, cal_coeffs):
-    t_cal_coeffs = np.float64(time_cal_coeffs)  # To be deleted later
-    t_ref = parse_time('1979-01-01', time=1)
+    t_cal_coeffs = []
+    t_cal_coeffs_flat = list(time_cal_coeffs.flatten())
```
Agreed

kakirastern commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,97 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        t_cal_coeffs.append(t_element)
+    t_cal_coeffs = np.array(t_cal_coeffs).reshape(5, 2)
+
+    for i in time_obs:
```
I will double check this...

kakirastern commented:
Found out that the problem stems from my new time conversions... So a test case for the first scenario `fit_iris_xput(time_obs, time_cal_coeffs0, cal_coeffs0)` with the data in [irispy/tests/test_iris_tools.py](https://github.com/sunpy/irispy/blob/5ae437568eaa78e85247ed6c0c398db2afd0f07c/irispy/tests/test_iris_tools.py) would be to look for the matrix

```
>>> t_diff
array([[  0.        ,  -0.24366872],
       [ -0.35318275,  -1.11430527],
       [ -1.11978097,  -1.19917864],
       [ -1.33880903,  -2.14373717],
       [ -2.14921287, -36.32580424]])
```
where every entry is in seconds...

kakirastern commented:
Turned out there are some `missing_axis` remaining in the files "spectrograph.py" and "test_sji.py". Have made the changes to `missing_axes` in commit 08d7cfc to debug...

kakirastern commented:
So locally everything seems fine except for the `missing_axes` bug... However, Travis is not happy with code snippets having the `parse_time` function in it, and is throwing some errors as the following:
```
______________ test_fit_iris_xput[input_arrays0-expected_array0] _______________
input_arrays = [array([1.0941696e+09]), array([[1.0941696e+09, 1.1018592e+09],
       [1.1053152e+09, 1.1293344e+09],
       [1.12950...
       [  0.22804098,   0.24711342,  -1.9952624 ],
       [  0.12897062,   0.19837108,  -0.8912509 ]], dtype=float32)]
expected_array = array([0.79865301])
    @pytest.mark.parametrize("input_arrays, expected_array", [
        ([time_obs, time_cal_coeffs0, cal_coeffs0], iris_fit_expected0),
        ([time_obs, time_cal_coeffs0, cal_coeffs1], iris_fit_expected1),
        ([time_obs, time_cal_coeffs0, cal_coeffs2], iris_fit_expected2),
        ([time_obs, time_cal_coeffs1, cal_coeffs3], iris_fit_expected3),
        ([time_obs, time_cal_coeffs1, cal_coeffs4], iris_fit_expected4),
        ([time_obs, time_cal_coeffs1, cal_coeffs5], iris_fit_expected5)])
    def test_fit_iris_xput(input_arrays, expected_array):
>       np_test.assert_almost_equal(fit_iris_xput(input_arrays[0], input_arrays[1], input_arrays[2]), expected_array, decimal=6)
irispy/tests/test_iris_tools.py:185: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
irispy/iris_tools.py:252: in fit_iris_xput
    time_obs = np.array([parse_time(float(t)) for t in time_obs])
irispy/iris_tools.py:252: in <listcomp>
    time_obs = np.array([parse_time(float(t)) for t in time_obs])
/home/travis/miniconda/envs/test/lib/python3.6/site-packages/sunpy/time/time.py:293: in parse_time
    rt = convert_time(time_string, format=format, **kwargs)
/home/travis/miniconda/envs/test/lib/python3.6/functools.py:807: in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
/home/travis/miniconda/envs/test/lib/python3.6/site-packages/sunpy/time/time.py:140: in convert_time
    return Time(time_string, format=format, **kwargs)
/home/travis/miniconda/envs/test/lib/python3.6/site-packages/astropy/time/core.py:399: in __init__
    precision, in_subfmt, out_subfmt)
/home/travis/miniconda/envs/test/lib/python3.6/site-packages/astropy/time/core.py:454: in _init_from_vals
    precision, in_subfmt, out_subfmt)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <[AttributeError("'Time' object has no attribute '_time'") raised in repr()] Time object at 0x7f9946ba7ba8>
val = array(1.0941696e+09), val2 = None, format = None, scale = None
precision = 3, in_subfmt = '*', out_subfmt = '*'
    def _get_time_fmt(self, val, val2, format, scale,
                      precision, in_subfmt, out_subfmt):
        """
        Given the supplied val, val2, format and scale try to instantiate
        the corresponding TimeFormat class to convert the input values into
        the internal jd1 and jd2.
    
        If format is `None` and the input is a string-type or object array then
        guess available formats and stop when one matches.
        """
    
        if format is None and val.dtype.kind in ('S', 'U', 'O', 'M'):
            formats = [(name, cls) for name, cls in self.FORMATS.items()
                       if issubclass(cls, TimeUnique)]
            err_msg = ('any of the formats where the format keyword is '
                       'optional {0}'.format([name for name, cls in formats]))
            # AstropyTime is a pseudo-format that isn't in the TIME_FORMATS registry,
            # but try to guess it at the end.
            formats.append(('astropy_time', TimeAstropyTime))
    
        elif not (isinstance(format, str) and
                  format.lower() in self.FORMATS):
            if format is None:
>               raise ValueError("No time format was given, and the input is "
                                 "not unique")
E               ValueError: No time format was given, and the input is not unique
```

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,99 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+    """
+    # Convert the observation time into a ``datetime.datetime`` object
+    # Then we would have "datetime.datetime(1970, 1, 1, 0, 0)"...
+    time_obs = np.array([parse_time(float(t), format = 'utime') for t in time_obs])
```
As @Cadair pointed out, this should be
```python
time_obs = parse_time(time_obs, format='utime')
```

kakirastern replied:
Yup, fixed that bit of the code...

kakirastern commented:
Reckon I can work more to polish up the documentation such as the docstrings where appropriate while we wait for the bugs to be fixed... might as well make myself useful to fill the time while we wait.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,102 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        raise ValueError("Incorrect number of elements either in time_cal_coeffs or in cal_coeffs.")
+
+    # Some time transformations.
+    # Convert the time_cal_coeffs given in the .geny file into a ``datetime.datetime`` object
```
This comment should actually say astropy Time now, right?

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,102 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        t_diff = [x.to(u.year) for x in t_diff]
+        
+        # To Convert to an array, quantities need to be dimensionless, hence dividing out the unit.
+        t_diff = np.array([x/u.year for x in t_diff])
```
This line ccan be combined with the one above by replacing the above with
```python
t_diff = np.array([x.to(u.year).value for x in t_diff])
```

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,102 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        aux_cal_coeffs = np.zeros(2*size_time_cal_coeffs[0])
+
+        # Exponent for transition between exp.decay intervals.
+        transition_exp = 1.5
```
No need to define this every time in the for loop.  Move to above the for loop

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,102 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        # Exponent for transition between exp.decay intervals.
+        transition_exp = 1.5
+
+        fit_out = np.zeros(len(list(time_obs)), dtype=np.float64)
```
Ah!  This is a bug.  I can only think it hasn't shown up because we haven't been inputing multiple ```time_obs``` in our tests.  This is to hold the fit values for each ```time_obs```.  But this for loop iterates through each ```time_obs```.  So ```fit_out``` is being erased each time.  THis is fine if you only go through the loop once but will cause problems if you go through more than once.

The solution is to delete this line and replace the [for loop definition](https://github.com/sunpy/irispy/pull/108/files#diff-dee23d5475a6ed5e368938455c92fb2eR270) with the following:
```python
fit_out = np.zeros(len(time_obs))
for i, t in enumerate(time_obs):
```

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,102 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+
+        # Looking for the closest time in the calibration time intervals.
+        # Differences are given in years before passing to the next stage.
+        t_diff = time_obs - t_cal_coeffs
```
Consistent with required change raised in above comment, this line would have to become
```python
t_diff = t - t_cal_coeffs
```

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,6 +220,102 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        # print(aux_cal_coeffs)
+        # print(cal_coeffs[:,:2].reshape((aux_cal_coeffs.shape[0])))
+        fit_out[count] = np.matmul(aux_cal_coeffs, cal_coeffs[:, :2].reshape((aux_cal_coeffs.shape[0])))
+        count += 1
```
Consistent with required change raised in above comment, delete ```count +=1``` and replace ```fit_out[count]``` with ```fit_out[i]```

kakirastern commented:
So a preliminary version of the modified `get_iris_response` function has been drafted, incorporating the time dependency in the determination of the effective areas based on the now debugged `fit_iris_xput` function.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -209,9 +216,199 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+            for k in range(len(iris_response["INDEX_EL_SJI"][j, :])):
+                area_pre_launch = area_pre_launch * np.iris_response["INDEX_EL_SJI"][j, k].T()
+                # Time dependent response
+                iris_fit_2 = fit_iris_xput(time_obs, iris_response["C_S_TIME"][j, :, :], iris_response["COEFFS_SJI"][j, :, :])
```
This should not be inside this for loop, as it is simply being overwritten each time.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -209,9 +216,199 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+            for k in range(len(iris_response["INDEX_EL_SJI"][nuv*2, :])):
+                area_sji = area_sji * np.iris_response["INDEX_EL_SJI"][nuv*2:(nuv*2)+1, k].T()
+            # Apply time dependent profile shape adjustment to FUV SJI
+            if -nuv:
```
The IDL line here is actually ```~nuv``` which means ```not nuv```.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -209,9 +216,199 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+                # Calculate baseline SG area for scaling purposes
+                area_sg = iris_response["GEOM_AREA"]
+                for k in range(len(iris_response["INDEX_EL_SG"][nuv, :])):
+                               area_sg = area_sg * np.iris_response["INDEX_EL_SG"][nuv, k].T()
```
See comment on analogous line above.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -672,15 +874,37 @@ def calculate_photons_per_sec_to_radiance_factor(
-    # Get effective area
-    ########### This needs to be generalized to the time of OBS once that functionality is written #########
+
+def _get_interpolated_effective_area(iris_fit, detector_type, obs_wavelength):
```
The API of this version of this function should not include ```iris_fit```, but simply the time at which the effective area is required:
```python
def _get_interpolated_effective_area(detector_type, obs_wavelength, **kwargs):
```
where the ```**kwargs``` get fed to ```get_iris_reponse```.

Since this function is used in a few places, in the future we could look for a way to make it more efficient by not calling ```get_iris_response``` but storing the response in the object somewhere.  For now though, go with the above API.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -694,11 +918,12 @@ def _get_interpolated_effective_area(detector_type, obs_wavelength):
     eff_area_interp_base_unit = u.Angstrom
     tck = interpolate.splrep(response_wavelength.to(eff_area_interp_base_unit).value,
-                             eff_area.to(eff_area_interp_base_unit ** 2).value, s=0)
+                             eff_area.to(eff_area_interp_base_unit ** 2).value * iris_fit, s=0)
```
This line should not be changed.  Instead this [line](https://github.com/sunpy/irispy/pull/108/files#diff-dee23d5475a6ed5e368938455c92fb2eR908) should be changed to
```python
iris_response = get_iris_response(**kwargs)
```
after the API has been changed in accordance with the above comment.

DanRyanIrish replied:
Delete ```* iris_fit```

DanRyanIrish commented:
Hi @kakirastern.  You've made some good progress here.  I've left some much overdue comments to get you moving again.

kakirastern commented:
Corrections for first review of rewritten `get_iris_response` done.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -212,104 +212,103 @@ def get_iris_response(time_obs, pre_launch=False, response_file=None, response_v
-                # Time dependent response
-                iris_fit_2 = fit_iris_xput(time_obs, iris_response["C_S_TIME"][j, :, :], iris_response["COEFFS_SJI"][j, :, :])
-            # Time dependent profiles
+    if iris_response["VERSION"] > 2:
```
```python
if iris_response["VERSION"] > 2 and time_obs is not None:
```

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -209,9 +216,199 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+            # Calculate pre-launch area from the individual elements
+            area_pre_launch = iris_response["GEOM_AREA"]
+            for k in range(len(iris_response["INDEX_EL_SJI"][j, :])):
+                area_pre_launch = area_pre_launch * np.iris_response["INDEX_EL_SJI"][j, k].T()
```
What is confusing about this line?  Here is some pseudo-code that might help.  The for loop should (hopefully) be able to be replaced by something like
```python
area_pre_launch = iris_response["GEOM_AREA"] * np.prod(iris_response["INDEX_EL_SJI"]["trans"], axis=correct_axis)
```

kakirastern commented:
Hi @DanRyanIrish I found out that some messages you composed on GitHub does not show on the page here, though I am getting them in the email, hence some confusion before. From now on I will be careful to double check both the conversation shown here and in the emails sent by GitHub...

kakirastern commented:
Hi @DanRyanIrish And I reason I will need to correct the bits of the code involving the [numpy.prod](https://docs.scipy.org/doc/numpy/reference/generated/numpy.prod.html)... I misinterpreted your usage of the correct_axis in the statement previously.

kakirastern commented:
Added test data for `get_iris_response` function.

DanRyanIrish commented on irispy/tests/test_spectrograph.py:
```diff
@@ -137,9 +137,9 @@ def test_fits_data_comparison(iris_l2_test_raster):
-    data3[hdulist[3].data == -200.] = np.nan
+    #data1[hdulist[1].data == -200.] = np.nan
+    #data2[hdulist[2].data == -200.] = np.nan
+    #data3[hdulist[3].data == -200.] = np.nan
```
These lines can just be deleted.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -137,11 +144,9 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
-                  "Version of this response file = {0}".format(response_version))
+            warnings.warn("Time-dependent effective area determination has been "
+                          "enabled in this version of IRISpy."
+                          "Version of this response file = {0}".format(response_version))
```
This warning can just be deleted.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -253,25 +258,25 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
             for k in range(len(iris_response["INDEX_EL_SJI"][j, :])):
-                area_pre_launch = iris_response["GEOM_AREA"] * np.prod(iris_response["INDEX_EL_SJI"][j, k]["trans"], axis=correct_axis)
+                #area_pre_launch = iris_response["GEOM_AREA"] * np.prod(iris_response["INDEX_EL_SJI"][j, k]["trans"], axis=correct_axis)
+                area_pre_launch = iris_response["GEOM_AREA"] * np.prod(iris_response["ELEMENTS"]["trans"][iris_response["INDEX_EL_SJI"]], axis=1)[j, k]
```
This should be moved to and replace the original definition [here](https://github.com/sunpy/irispy/pull/108/files/fe1f1daf7d3d4f7d8333a2c9961960d7300a5f64..e43a794df5b3b9b2a30298a2f6f989a6fa3437f7#diff-dee23d5475a6ed5e368938455c92fb2eR263)

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -282,7 +287,8 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
                 for k in range(len(iris_response["INDEX_EL_SG"][nuv, :])):
-                    area_sg = iris_response["GEOM_AREA"] * np.prod(iris_response["INDEX_EL_SG"][nuv, k]["trans"], axis=correct_axis)
+                    #area_sg = iris_response["GEOM_AREA"] * np.prod(iris_response["INDEX_EL_SG"][nuv, k]["trans"], axis=correct_axis)
+                    area_sg = iris_response["GEOM_AREA"] * np.prod(iris_response["ELEMENTS"]["trans"][iris_response["INDEX_EL_SJI"]], axis=1)[nuv, k]
```
This should be moved to and replace the original definition [here](https://github.com/sunpy/irispy/pull/108/files/fe1f1daf7d3d4f7d8333a2c9961960d7300a5f64..e43a794df5b3b9b2a30298a2f6f989a6fa3437f7#diff-dee23d5475a6ed5e368938455c92fb2eR288) and the for loop deleted.

kakirastern commented:
Just checked that after getting the `get_iris_response` function working, have successfully got the second halves of the `AREA_SG` and `AREA_SJI` to agree with the IDL data provided in the [.sav file](https://github.com/sunpy/irispy/blob/ee4ee291d9f7f82127b197c029177ba0676d291c/irispy/data/idl_iris_get_response_20130903_new.sav) working. Will need to work out a strategy to debug the function further.

kakirastern commented:
Fixed the bug in the `AREA_SG` computation. Now down to the `AREA_SJI` bugs in Part 3 of the code as indicated in the code base or `iris_tools` module.

hayesla commented on irispy/iris_tools.py:
```diff
@@ -165,53 +166,247 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+            iris_response["AREA_SG"][1, w_nuv] = interpol_nuv(iris_response["LAMBDA"][w_nuv])
+        
+        # 3. SJI effective areas
+        shp_sji = iris_response["COEFFS_SJI"].shape
```
this should only be for version >=003
so need condition:
`if int(iris_response['VERSION']) >= 003:`

hayesla commented on irispy/iris_tools.py:
```diff
@@ -165,53 +166,247 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        # 3. SJI effective areas
+        shp_sji = iris_response["COEFFS_SJI"].shape
+        area_pre_launch = iris_response["GEOM_AREA"]
+        for j in range(shp_sji[0]):
```
for j in range(0, shp_sji[0]):
	area_pre_launch = iris_response['GEOM_AREA']
	for k in range(0, len(iris_response['INDEX_EL_SJI'][j,:])):
		index_value = iris_response['INDEX_EL_SJI'][j, k]
		area_pre_launch = area_pre_launch * iris_response['ELEMENTS'].trans[index_value]

hayesla commented on irispy/iris_tools.py:
```diff
@@ -165,53 +166,247 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        area_pre_launch = iris_response["GEOM_AREA"]
+        for j in range(shp_sji[0]):
+            # Calculate pre-launch area from the individual elements
+            area_pre_launch = iris_response["GEOM_AREA"] * np.prod(iris_response["ELEMENTS"]["trans"][iris_response["INDEX_EL_SJI"]], axis=1)
```
get rid of this

hayesla commented on irispy/iris_tools.py:
```diff
@@ -165,53 +166,247 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+            # Time dependent profiles
+            for k in range(n_time_obs):
+                iris_response["AREA_SJI"][j, :] = area_pre_launch[j] * iris_fit_sji[k]
+            area_sji = iris_response["GEOM_AREA"]
```
get rid of this

hayesla commented on irispy/iris_tools.py:
```diff
@@ -165,53 +166,247 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+            for k in range(n_time_obs):
+                iris_response["AREA_SJI"][j, :] = area_pre_launch[j] * iris_fit_sji[k]
+            area_sji = iris_response["GEOM_AREA"]
+            for nuv in range(2):
```
now this is where the condition starts for versions greater than 3 so need
`elif int(iris_response['VERSION']) >3:`

kakirastern commented:
So now `response_version=1, 2` are working perfectly. Except for versions 3, 4, which are not yielding the desired outputs.

kakirastern commented:
Both `get_iris_response` and `fit_iris_xput` functions working now and are generating outputs as expected. Tests have been written to check both are performing well. (I have set the number of decimal places for the agreement to `6` for now.) Once everything checks out I can then move on to the final stage of the project...

kakirastern commented:
I believe the next stage is this: 
> Updated intensity conversion methods between instrument and physical units that correct for the time observations were taken.

But I will need guidance to do that.

hayesla commented on irispy/iris_tools.py:
```diff
@@ -66,18 +73,21 @@
-def get_iris_response(pre_launch=False, response_file=None, response_version=None,
-                      force_download=False):
+
+def get_iris_response(time_obs=None, pre_launch=False, response_file=None, response_version=None,
```
I think that if nothing is given then it should default to version 4. Such that you can run `get_iris_response()` it should run for version 4.

kakirastern replied:
Maybe another PR for this?

hayesla commented on irispy/iris_tools.py:
```diff
@@ -66,18 +73,21 @@
-def get_iris_response(pre_launch=False, response_file=None, response_version=None,
-                      force_download=False):
+
+def get_iris_response(time_obs=None, pre_launch=False, response_file=None, response_version=None,
```
Theres a bug here if someone gives a `time_obs = parse_time('2013-09-03')`. or if its not in an array format. I think you should have this as a check. Also you should make sure that its a time object - currently it will do something with a string which is buggy.

hayesla commented on irispy/tests/test_iris_tools.py:
```diff
@@ -119,32 +150,164 @@ def test_uncalculate_exposure_time_correction(input_arrays, old_unit, exposure_t
 
-def test_get_iris_response():
-    pass
+# Tests for get_iris_response function
```
theses tests can be tidied up but for now its grand

hayesla commented on irispy/iris_tools.py:
```diff
@@ -98,15 +109,14 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
         comment: `str`
         version: `int`
-        version_date: `datetime.datetime`
+        version_date: `time.time`
```
`astropy.time.Time`

DanRyanIrish replied:
Yes.  All instances of ```"`time.time`"``` should be changed to ```"`astropy.time.Time`"```

hayesla commented on irispy/iris_tools.py:
```diff
@@ -87,8 +97,9 @@ def get_iris_response(pre_launch=False, response_file=None, response_version=Non
+        Various parameters regarding IRIS response or effective area structure.
+        Includes the following keys:
+        date_obs: `time.time`
         lambda: `astropy.units.Quantity`
```
I dont know how we feel about all the output keywords being capitalized ... something to talk about in the future after this PR

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -672,11 +890,13 @@ def calculate_photons_per_sec_to_radiance_factor(
+    return constants.h * constants.c / wavelength / u.photon / spectral_dispersion_per_pixel / eff_area_interp / solid_angle
+
 
 def _get_interpolated_effective_area(detector_type, obs_wavelength):
```
Too much has been reverted in a recent commit.  Change lines 901 to 904 to:
```python
def _get_interpolated_effective_area(detector_type, obs_wavelength, **kwargs):
   """Put docstring back"""
    # Get effective area
    iris_response = get_iris_response(pre_launch=True)
```
The rest of the function remains untouched.

kakirastern replied:
Yup, done. Will work on an improvement of this docstring in the next PR... Will need to go through each and every docstring then update and expand on it where appropriate.

kakirastern commented:
@DanRyanIrish and @hayesla I have resolved all the issues brought up in the reviews and have resolved them one-by-one. Hopefully this PR can be merged soon so new PR's can be open to make it complete.

DanRyanIrish commented:
Hi @kakirastern.  With @hayesla's help I figured out the confusion we were having earlier that prevented us from merging this PR today.  I didn't realise that ```iris_fit``` was a required input to ```_get_interpolated_effective_area``` because it wasn't in the ```def``` statement.

So I suggest that one last change be made: undo the changes in commit "Revert changes previously made to the "_get_interpolated_effective_area" method"``` (c3abaf7524359da41f94be2418fb58a402b43d92).  Then we will merge.  Apologies for the confusion!

kakirastern commented:
@DanRyanIrish No worries. Also, will I need to add `iris_fit` as an argument to `_get_interpolated_effective_area`? Or would adding `**kwargs` suffice?

kakirastern commented:
Rebased PR and dropped commit c3abaf7524359da41f94be2418fb58a402b43d92. Then added another commit to change `time.time` in the in-code documentation to `astropy.time.Time` as previously instructed.

hayesla commented on irispy/iris_tools.py:
```diff
@@ -672,16 +890,38 @@ def calculate_photons_per_sec_to_radiance_factor(
-    # Get effective area
-    ########### This needs to be generalized to the time of OBS once that functionality is written #########
-    iris_response = get_iris_response(pre_launch=True)
+def _get_interpolated_effective_area(detector_type, obs_wavelength, *args, **kwargs):
```
@kakirastern  you should have iris_fit here as an argument

kakirastern replied:
Done

DanRyanIrish commented:
Hi @kakirastern.  I see that you've made those changes.  Once the CI passes I'm happy to merge this.  You've done a great job on a very tricky assignment!  How about you @hayesla?  Are you happy to approve this?

hayesla commented:
Good with me!
